### PR TITLE
ipc4: Correction of component id printed in tr_err

### PR DESCRIPTION
**Pull request by softwarecki**
Corrected component id printed in tr_err.
Added some important comments.
Added asserts witch helped me debug the code.

### CODE REVIEW

**Comment by sofci** [comment]
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by mwasko** [comment]
> 

test this please

File: src/include/sof/audio/audio_stream.h [hunk 1]
```diff
@@ -228,6 +228,8 @@ static inline void *audio_stream_wrap(const struct audio_stream *buffer,
 		ptr = (char *)buffer->addr +
 			((char *)ptr - (char *)buffer->end_addr);
 
+	assert((intptr_t)ptr <= (intptr_t)buffer->end_addr);
```
Comment (lgirdwood) [inline src/include/sof/audio/audio_stream.h:231]: Lets log this as an error instead as it unrecoverable if deployed in production.

File: src/audio/copier.c [hunk 2]
```diff
@@ -46,6 +46,9 @@ DECLARE_SOF_RT_UUID("copier", copier_comp_uuid, 0x9ba00c83, 0xca12, 0x4a83,
 DECLARE_TR_CTX(copier_comp_tr, SOF_UUID(copier_comp_uuid), LOG_LEVEL_INFO);
 
 struct copier_data {
+	/* Must be the 1st field, function ipc4_create_buffer casts components private data
+	 * as ipc4_base_module_cfg!
+	 */
```
Comment (ranj063) [inline src/audio/copier.c:51]: Separate commit please

File: src/include/sof/audio/audio_stream.h [hunk 3]
```diff
@@ -525,6 +527,7 @@ static inline int
 audio_stream_bytes_without_wrap(const struct audio_stream *source,
 				const void *ptr)
 {
+	assert((intptr_t)source->end_addr >= (intptr_t)ptr);
```
Comment (ranj063) [inline src/include/sof/audio/audio_stream.h:530]: These should be separate commits too. Preferably errors

File: src/include/sof/audio/audio_stream.h (hunk 1 above)
Comment (lgirdwood) [reply src/include/sof/audio/audio_stream.h:231]: ping - we need to be more recoverable when we fail. e.g. so we can attempt graceful powerdown and reboot.

**Review by lgirdwood — COMMENTED** [review]
ping ?

File: src/include/sof/audio/audio_stream.h (hunk 1 above)
Comment (lgirdwood) [reply src/include/sof/audio/audio_stream.h:231]: I know assert() is in other places, but the intention is to remove as we go along - we should only really be asserting if its likely impossible to gracefully recover.

**Comment by softwarecki** [comment]
Do we have any functions that will perform graceful power down and reboot, which I could use?

**Comment by lgirdwood** [comment]
> Do we have any functions that will perform graceful power down and reboot, which I could use?

No not currently, we just have the IPC method to cleanly shutdown. All we can do now is for the error to propagate up the stack to the host and the host can then decide what action to take. So in this instance we could signal an error and have the pipeline stop (which may require other logic).

If we can cleanly shutdown then we can avoid situations where the DMA may still be active. I know this is not the aim of this PR, but we need to think about a clean shutdown path as a separate feature for future work.

**Comment by softwarecki** [comment]
Suppose the functions return an error (-1 / NULL) when they detect a problem with the buffer. The most of functions, which call `audio_stream_bytes_without_wrap` or `audio_stream_wrap` returns void. At this moment we can't propagate detected error. Currently we have 2 options: assert and hang on error or return invalid value which will lead to overwrite a random memory location.

**Comment by lgirdwood** [comment]
> Suppose the functions return an error (-1 / NULL) when they detect a problem with the buffer. The most of functions, which call `audio_stream_bytes_without_wrap` or `audio_stream_wrap` returns void. At this moment we can't propagate detected error. Currently we have 2 options: assert and hang on error or return invalid value which will lead to overwrite a random memory location.

Yep, it's going to be difficult. Lets take this atm as this cant be recovered from, we can re-look at how to deal with fatal errors in a ways that allows the host to cleanly recover.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@softwarecki looks like we have a conflict, can you fix. Thanks!

File: src/include/sof/audio/audio_stream.h (hunk 3 above)
Comment (marc-hb) [reply src/include/sof/audio/audio_stream.h:530]: Pretty sure these asserts are buggy, see demo in https://github.com/thesofproject/sof/pull/7929#discussion_r1264423943
